src/icrar/leap-accelerate/common/vector_extensions.h

```diff
@@ -23,17 +23,68 @@
 
 #pragma once
 
+#include <iostream>
 #include <vector>
+#include <functional>
 
-template<typename T>
-bool isApprox(const std::vector<T>& lhs, const std::vector<T>& rhs, T threshold)
+template <typename T>
+std::ostream& operator<<(std::ostream& os, const std::vector<T>& v)
 {
-    for(size_t i = 0; i < lhs.size(); ++i)
+    os << "{";
+    for (int i = 0; i < v.size(); ++i)
+    { 
+        os << v[i]; 
+        if (i != v.size() - 1) 
+            os << ", "; 
+    } 
+    os << "}\n"; 
+    return os;
+}
+
+namespace icrar
+{
+    /**
+     * @brief Returns of true if all vector elements of @param lhs are within the threshold difference to @param rhs 
+     * 
+     * @tparam T 
+     * @param lhs 
+     * @param rhs 
+     * @param threshold 
+     * @return true 
+     * @return false 
+     */
+    template<typename T>
+    bool isApprox(const std::vector<T>& lhs, const std::vector<T>& rhs, T threshold)
     {
-        if(std::abs(lhs[i] - rhs[i]) >= threshold)
+        if(lhs.size() != rhs.size())
         {
             return false;
         }
+        
+        for(size_t i = 0; i < lhs.size(); ++i)
+        {
+            if(std::abs(lhs[i] - rhs[i]) >= threshold)
+            {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * @brief Performs a std::transform on a newly allocated std::vector 
+     * 
+     * @tparam T The input vector type
+     * @tparam R the output vector type
+     * @param vector 
+     * @return std::vector<R> 
+     */
+    template<typename R, typename T>
+    std::vector<R> vector_map(const std::vector<T>& vector, std::function<R(const T&)> lambda)
```

Comment:
That sounds like a good candidate for an sfinae static assertion

```
    template<typename R, typename T, typename Op>
    std::vector<R> vector_map(const std::vector<T>& vector, Op lambda)
    {
        static_assert(std::is_assignable<std::function<R(const T&)>, Op>::value, "lambda argument must be a function of signature R(const T&)");
    }
```

Reply:
I was thinking more on using `std::result_of` to avoid having to pass `R` and instead deduce it from `Op`. Something like:

```c++
template <typename T, typename Op>
std::vector<typename std::result_of<Op(T)>::type> vector_map(const std::vector<T>& vector, Op &&f)
{
    using R = typename std::result_of<Op(T)>::type;
    std::vector<R> results(vector.size());
    std::transform(vector.begin(), vector.end(), results.begin(), f);
    return results;
}
```

Then from the caller's point of view both `T` and `Op` are deduced from the arguments, so there's no need to explicitly give any template parameters when making the call.